### PR TITLE
Knip CI

### PR DESCRIPTION
**Pull request by gitcoindev**
Resolves [#897](https://github.com/ubiquity/ubiquibot/issues/897)

Quality Assurance: 

a) Annotations: https://github.com/gitcoindev/ubiquibot/pull/17/files
b) PR comment: https://github.com/gitcoindev/ubiquibot/pull/17#issuecomment-1828200946

Screenshot:

Screenshot from 2023-11-28 12-23-08

### CODE REVIEW

File: .github/workflows/knip.yml
Comment (0x4007) [inline .github/workflows/knip.yml]: Where can I see this source code?

File: .github/workflows/knip.yml
Comment (0x4007) [inline .github/workflows/knip.yml]: We should use latest LTS

**Review by 0x4007 — APPROVED** [review]

File: .github/workflows/knip.yml
Comment (gitcoindev) [reply .github/workflows/knip.yml]: @pavlovcik it is available here: https://github.com/gitcoindev/knip-reporter

File: .github/workflows/knip.yml
Comment (gitcoindev) [reply .github/workflows/knip.yml]: good point, I will check with the latest LTS as well

Comment (0x4007) [reply .github/workflows/knip.yml]: Is this finished? Should I merge?

File: .github/workflows/knip.yml
Comment (0x4007) [reply .github/workflows/knip.yml]: Perhaps its best to place this under Ubiquity? Looks like they updated the code too.

File: .github/workflows/knip.yml
Comment (gitcoindev) [reply .github/workflows/knip.yml]: @pavlovcik I will test with latest LTS today and let you know asap.

File: .github/workflows/knip.yml
Comment (gitcoindev) [reply .github/workflows/knip.yml]: Ok let me check the update.

File: .github/workflows/knip.yml
Comment (gitcoindev) [reply .github/workflows/knip.yml]: I rebased to latest knip v3 support in my fork, fixed the tests and opened a PR upstream https://github.com/Codex-/knip-reporter/pull/15 will try to push this, if this is accepted we will be able to use an official release.

File: .github/workflows/knip.yml
Comment (gitcoindev) [reply .github/workflows/knip.yml]: Let's wait a few days for the upstream pull request to be merged.  If it will stall we'll move under Ubiquity, but since the author released knip-reporter with v3 support this week there is a high chance he will pick it up soon.

Comment (0x4007) [reply .github/workflows/knip.yml]: Okay keep me posted!

File: .github/workflows/knip.yml
Comment (gitcoindev) [reply .github/workflows/knip.yml]: My pull request was extended and approved, merged upstream. I will now verify this on my fork and let you know if this works as expected.

Comment (gitcoindev) [reply .github/workflows/knip.yml]: I spent quite some time on debugging and asked the author why it does not work as expected on latest upstream release https://github.com/Codex-/knip-reporter/issues/9#issuecomment-1853771495 but does work on my fork let's wait for his response.

Comment (gitcoindev) [reply .github/workflows/knip.yml]: @pavlovcik after a few hours I found the culprit and solved the problem.

knip v3 does not work well with typescript < 5 , does not throw errors but its output json does not contain all needed data. I had to update ubiquibot's dev dependencies to typescript ^5.0.4 and jest + ts-jest to v29 and it works like a charm, annotations in action are visible in my qa at https://github.com/gitcoindev/ubiquibot/pull/17/files

It is a bit late for me today, I will check tomorrow if this did not break anything and will add the typescript and jest upgrade as well to refactor/general

**Comment by 0x4007** [comment]
CI failed. What's the status on this?

**Comment by gitcoindev** [comment]
> CI failed. What's the status on this?

Permissions issue, I will try to reproduce on my org repo as on my fork it looks fine.

**Comment by gitcoindev** [comment]
> > CI failed. What's the status on this?
> 
> Permissions issue, I will try to reproduce on my org repo as on my fork it looks fine.

@pavlovcik  I think I found the root cause. 

https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#changing-the-permissions-in-a-forked-repository

I suspect that your fork's https://github.com/pavlovcik/ubiquibot settings restrict access for my pull requests  to add / modify comments and would have to change 'Workflow permissions' settings

Screenshot from 2023-12-19 22-46-41


https://docs.github.com/en/actions/security-guides/automatic-token-authentication#modifying-the-permissions-for-the-github_token

I will check if perhaps also permissions can be reduced, currently the yaml workflow file uses `permissions: write-all` which seems too much.

**Comment by 0x4007** [comment]
Okay I'll try later today

**Comment by gitcoindev** [comment]
Hi @pavlovcik I dug deeper to check why CI fails here. It fails because the pull request is opened from a fork to a fork and not to the original bot repository. This situation is described at https://github.blog/2020-08-03-github-actions-improvements-for-fork-and-pull-request-workflows/#improvements-for-public-repository-forks and https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks . 

I was able to reproduce the same permission issue in my personal fork https://github.com/gitcoindev/ubiquibot vs my org fork https://github.com/korrrba/ubiquibot as well. Therefore, once the pull request is merged into the original bot repository, the knip ci workflow will work correctly.

**Pull request merged**
